frontend/src/components/EmbeddedApp.tsx

```diff
@@ -0,0 +1,201 @@
+import {
+  AppIFrameMessage,
+  AppIFrameRequest,
+  AppIFrameResponse,
+} from "@/types/app-iframe-message";
+import styled from "@emotion/styled";
+import {
+  forwardRef,
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+} from "react";
+
+const StyledIFrame = styled.iframe`
+  width: 100%;
+  border: none;
+
+  flex-grow: 1;
+`;
+
+const MAX_COUNTER = 1000000;
+let counter = 0;
+
+const postMessageToIFrame = (
+  iFrame: HTMLIFrameElement,
+  message: AppIFrameMessage,
+) => {
+  if (!iFrame.contentWindow) {
+    console.error(
+      "Cannot post message to iframe without content window:",
+      iFrame,
+    );
+    return;
+  }
+
+  // get target origin from the iframe's src attribute
+  const targetOrigin = new URL(iFrame.src).origin;
+
+  iFrame.contentWindow.postMessage(message, {
+    targetOrigin,
+  });
+};
+
+export interface EmbeddedAppRef {
+  sendRequest<ProcedureName extends AppIFrameRequest["procedure"]>(
+    request: Omit<AppIFrameRequest, "id" | "type"> & {
+      procedure: ProcedureName;
+    },
+  ): Promise<AppIFrameResponse & { procedure: ProcedureName }>;
+}
+
+export interface Props {
+  src: string;
+}
+
+const EmbeddedApp = forwardRef<EmbeddedAppRef, Props>(function EmbeddedApp(
+  { src },
+  ref,
+) {
+  const [iFrame, setIFrame] = useState<HTMLIFrameElement | null>(null);
```

Comment:
```suggestion
  const [iframe, setIFrame] = useState<HTMLIFrameElement | null>(null);
```

When not capitalized, it's usually spelled `iframe` - I'd try to rename `iFrame` to `iframe` everywhere in this file.